inst/include/cpp11/r_string.hpp

```diff
@@ -16,6 +16,9 @@ class r_string {
   r_string(const char* data) : data_(safe[Rf_mkCharCE](data, CE_UTF8)) {}
   r_string(const std::string& data) : data_(safe[Rf_mkCharCE](data.c_str(), CE_UTF8)) {}
 
+  template <int N>
+  r_string(char data [N]) : data_(safe[Rf_mkCharLenCE](data, N, CE_UTF8)) {}
```

Comment:
```suggestion
  r_string(char (&data)[N]) : data_(safe[Rf_mkCharLenCE](data, N, CE_UTF8)) {}
```

Reply:
More to the point, I don't see a definition for `r_vector<T>::proxy::operator=` [inline](https://github.com/r-lib/cpp11/blob/834e74e82b2c2731cae573133c26446e56a39bf5/inst/include/cpp11/r_vector.hpp#L226-L227) or out of line with the other operator defs [where I'd expect](https://github.com/r-lib/cpp11/blob/834e74e82b2c2731cae573133c26446e56a39bf5/inst/include/cpp11/r_vector.hpp#L874). Same for `proxy<T>::operator T`. I'd expect you to need something like

```diff
diff --git a/inst/include/cpp11/r_vector.hpp b/inst/include/cpp11/r_vector.hpp
index 1222af4..6e4a203 100644
--- a/inst/include/cpp11/r_vector.hpp
+++ b/inst/include/cpp11/r_vector.hpp
@@ -872,6 +872,18 @@ inline r_vector<T>::operator SEXP() const {
   return data_;
 }
 
+template <typename T>
+template <typename U>
+typename r_vector<T>::proxy& r_vector<T>::proxy::operator=(const U& rhs) {
+  set_value_at(data_, index_, is_altrep_, as_sexp(rhs));
+  return *this;
+}
+
+template <typename T>
+r_vector<T>::proxy::operator T() const {
+  return as_cpp<T>(get_value_at(data_, index_, is_altrep_));
+}
+
 template <typename T>
 inline typename r_vector<T>::proxy& r_vector<T>::proxy::operator+=(const T& rhs) {
   operator=(static_cast<T>(*this) + rhs);
```

Reply:
@jimhester I invented set_value_at and get_value_at, what should those be replaced with?

Reply:
They are defined in the headers the implement the specializations, e.g. https://github.com/r-lib/cpp11/blob/c17cd96188734bf5ef9375088e6706bdc2933f31/inst/include/cpp11/doubles.hpp#L51-L72

Reply:
Hmmm, defining them this way opens us up to confusing errors. We won't get a compilation error for
```c++
writable::doubles d = //...
d[3] = "wtf";
```
instead we'll get a linker error complaining that `r_vector<double>::proxy& r_vector<double>::proxy::operator=<char(&)[4]>(char (&)[4])` is not defined. What would you think of moving to a mixin based impl for vector proxies? Sketch:

https://github.com/r-lib/cpp11/compare/master...bkietz:r_vector_proxy-mixins

Reply:
I think I would rather revert https://github.com/r-lib/cpp11/pull/57 make users use `TRUE` and `FALSE` as it was previously then complicate this code further.